ci/cloudbuild/notifiers/alerts/function/function.cc

```diff
@@ -0,0 +1,86 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <google/cloud/functions/cloud_event.h>
+#include <cppcodec/base64_rfc4648.hpp>
+#include <curl/curl.h>
+#include <fmt/core.h>
+#include <nlohmann/json.hpp>
+#include <cctype>
+#include <iostream>
+#include <memory>
+#include <regex>
+#include <sstream>
+#include <stdexcept>
+
+namespace {
+
+void LogError(std::string const& msg) {
+  auto const json = nlohmann::json{{"severity", "error"}, {"message", msg}};
+  std::cerr << json.dump() << "\n";
+}
+
+auto MakeChatPayload(nlohmann::json const& build) {
+  auto const trigger_name = build["substitutions"].value("TRIGGER_NAME", "");
+  auto const log_url = build.value("logUrl", "");
+  auto text = fmt::format(
+      R""(Failed build: *{trigger_name}* (<{log_url}|Build Log>))"",
+      fmt::arg("trigger_name", trigger_name), fmt::arg("log_url", log_url));
+  return nlohmann::json{{"text", std::move(text)}};
+}
+
+void HttpPost(std::string const& url, std::string const& data) {
+  static constexpr auto kContentType = "Content-Type: application/json";
+  using Headers = std::unique_ptr<curl_slist, decltype(&curl_slist_free_all)>;
+  auto const headers = std::make_unique<Headers>(
```

Comment:
Is this creating a `std::unique_ptr<std::unique_ptr<curl_slist, decltype(&curl_list_free_all)>>` ?

Reply:
:facepalm: